File: app/src/main/java/com/hitstudio/apps/netbook/ui/viewmodel/NoteViewModel.java
package com.hitstudio.apps.netbook.ui.viewmodel;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.ViewModel;

import com.hitstudio.apps.netbook.domain.model.Note;
import com.hitstudio.apps.netbook.domain.repository.NoteRepository;

import java.util.List;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

import javax.inject.Inject;

import dagger.hilt.android.lifecycle.HiltViewModel;

@HiltViewModel
public final class NoteViewModel extends ViewModel {
    private final NoteRepository noteRepository;
    private final LiveData<List<Note>> notes;
    private final ExecutorService databaseExecutor = Executors.newSingleThreadExecutor();

    @Inject
    public NoteViewModel(NoteRepository noteRepository) {
        this.noteRepository = noteRepository;
        this.notes = noteRepository.getAllNotes();
    }

    public LiveData<List<Note>> getNotes() {
        return notes;
    }

    public void deleteNote(Note note) {
        databaseExecutor.execute(() -> noteRepository.deleteNote(note));
    }

    @Override
    protected void onCleared() {
        databaseExecutor.shutdown();
    }
}

File: app/src/main/java/com/hitstudio/apps/sharednotebook/ui/viewmodel/NoteViewModel.kt
package com.hitstudio.apps.sharednotebook.ui.viewmodel

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.hitstudio.apps.sharednotebook.domain.model.Note
import com.hitstudio.apps.sharednotebook.domain.repository.NoteRepository
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.launch
import javax.inject.Inject

@HiltViewModel
class NoteViewModel @Inject constructor(
    private val noteRepository: NoteRepository
) : ViewModel() {

    val notes: StateFlow<List<Note>> = noteRepository.getAllNotes()
        .stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000), emptyList())

    fun deleteNote(note: Note) {
        viewModelScope.launch {
            noteRepository.deleteNote(note)
        }
    }
}